source/common/shared_pool/shared_pool.h

```diff
@@ -30,60 +31,63 @@ namespace SharedPool {
  * There is also a need to ensure that the thread where ObjectSharedPool's destructor is also in the
  * main thread, or that ObjectSharedPool destruct before the program exit
  */
-template <typename T, typename HashFunc = std::hash<T>,
+template <typename T, typename HashFunc = std::hash<T>, typename EqualFunc = std::equal_to<T>,
           class = typename std::enable_if<std::is_copy_constructible<T>::value>::type>
-class ObjectSharedPool : public Singleton::Instance,
-                         public std::enable_shared_from_this<ObjectSharedPool<T, HashFunc>>,
-                         NonCopyable {
+class ObjectSharedPool
+    : public Singleton::Instance,
+      public std::enable_shared_from_this<ObjectSharedPool<T, HashFunc, EqualFunc>>,
+      NonCopyable {
 public:
   ObjectSharedPool(Event::Dispatcher& dispatcher)
       : thread_id_(std::this_thread::get_id()), dispatcher_(dispatcher) {}
 
-  void deleteObject(const size_t hash_key) {
+  void deleteObject(T* ptr) {
     if (std::this_thread::get_id() == thread_id_) {
-      // There may be new inserts with the same hash value before deleting the old element,
-      // so there is no need to delete it at this time.
-      if (object_pool_.find(hash_key) != object_pool_.end() &&
-          object_pool_[hash_key].use_count() == 0) {
-        object_pool_.erase(hash_key);
+      if (auto iter = object_pool_.find(ptr); iter != object_pool_.end()) {
+        // It is possible that the entry in object_pool_ corresponds to a
+        // different weak_ptr, due to a race condition in a shared_ptr being
+        // destroyed on another thread, and getObject() being called on the main
+        // thread.
+        if (iter->use_count() == 0) {
+          object_pool_.erase(iter);
+        }
       }
+      // Wait till here to delete the pointer because we don't want the OS to
+      // reallocate the memory location before this method completes to prevent
+      // "hash collisions".
+      delete ptr;
     } else {
       // Most of the time, the object's destructor occurs in the main thread, but with some
       // exceptions, it is destructed in the worker thread. In order to keep the object_pool_ thread
       // safe, the deleteObject needs to be delivered to the main thread.
       auto this_shared_ptr = this->shared_from_this();
       // Used for testing to simulate some race condition scenarios
       sync_.syncPoint(DeleteObjectOnMainThread);
-      dispatcher_.post([hash_key, this_shared_ptr] { this_shared_ptr->deleteObject(hash_key); });
+      dispatcher_.post([ptr, this_shared_ptr] { this_shared_ptr->deleteObject(ptr); });
```

Comment:
WDYT of having breaking out a separate function deleteObjectMainThread where we put lines 46-58, and having the post call that directly, rather than re-checking the thread ID?

It might make the intent clearer.

Reply:
Sounds good. Also made the methods private.